File: scripts/linters/no_format_trailing_comma.rb
def no_format_trailing_comma(root_dir, excludes = [])
  crates_dir = root_dir.join('crates')
  raw = `grep -rn --include='*.rs' -e ',)' #{crates_dir}`

  errors = []
  raw.each_line do |line|
    line = line.chomp
    path, lineno, content = line.split(':', 3)
    next if content.nil?

    relative = Pathname.new(path).relative_path_from(root_dir).to_s
    next if excludes.include?(relative)

    # `(,)` is a macro repetition fragment (e.g. `$(,)?`).
    next if content.match?(/\(,\)/)

    errors << "#{relative}:#{lineno}: trailing `,)` before a closing paren"
  end

  return true if errors.empty?

  puts 'Found `,)` introduced by formatting. Remove it.'
  errors.each do |err|
    puts "  #{err}"
  end
  false
end

File: scripts/linters/src/Linters/NoFormatTrailingComma.php
<?php

declare(strict_types=1);

namespace Shirabe\Lint\Linters;

use Shirabe\Lint\Linter;
use Shirabe\Lint\Support\FileFinder;
use Shirabe\Lint\Support\Paths;

final class NoFormatTrailingComma implements Linter
{
    public function name(): string
    {
        return 'no_format_trailing_comma';
    }

    public function failureIntro(): string
    {
        return 'Found `,)` introduced by formatting. Remove it.';
    }

    public function check(string $rootDir, array $excludes): array
    {
        $errors = [];

        foreach (FileFinder::rustFiles($rootDir) as $path) {
            $relative = Paths::relativeTo($rootDir, $path);
            if (in_array($relative, $excludes, true)) {
                continue;
            }

            foreach (file($path) as $idx => $raw) {
                if (!str_contains($raw, ',)')) {
                    continue;
                }
                // `(,)` is a macro repetition fragment (e.g. `$(,)?`).
                if (str_contains($raw, '(,)')) {
                    continue;
                }

                $errors[] = "{$relative}:" . ($idx + 1) . ': trailing `,)` before a closing paren';
            }
        }

        return $errors;
    }
}
